### ISSUE TITLE
Why is the 'results' array required?

### ISSUE THREAD

**Issue by michaelcfanning** [oasis-tcs/sarif-spec#232]
The spec says that the 'results' array is required. I'm not sure this should be the case. For example, if a tool cannot initialize to any degree, this might be expressed as a tool notification (and no results array would exist). You can imagine a similar case for configuration notifications.

It is helpful to provide some guidance around what constitutes a successful/complete analysis. The absence of a results array is a good indicator that the run itself is compromised.

**Comment by michaelcfanning** [comment]
Notes that "results" was not required in SARIF v1. I can't find an issue in history that shows when this was reconsidered (but haven't mined everything I could to uncover this point)

**Comment by lgolding** [comment]
@michaelcfanning

We required `run.results` so a tool could give a positive indication of the condition "I performed the analysis, but there weren't any results" by emitting an empty array. We did not consider your scenario, where it makes sense for a tool to omit even the empty array to convey the message "I couldn't even perform the analysis."

The _schema_ does not specify `run.results` required because we wanted to use the same schema for a normal tool output file as for a file that contains only rule metadata or other resources. In that case, `run.results` doesn't make sense.

This is all exactly as it was in v1: the spec [required it](https://rawgit.com/sarif-standard/sarif-spec/master/Static%20Analysis%20Results%20Interchange%20Format%20(SARIF).html#run-results):

> If the analysis tool was run with the intent of scanning files and producing results, then the `run` object shall contain a property named `results` whose value is an array containing zero or more unique (§5.9) `result` objects (§5.17), each of which represents a single result detected in the course of the run.

... but the schema did not.

**Comment by lgolding** [comment]
@michaelcfanning 

TL;DR: We should add text as you suggest, explaining that you should omit `run.results` if the tool was unable even to begin its analysis.

**Comment by michaelcfanning** [comment]
The V2 schema contains the following, which, according to your last post, is a bug. yes? 

      "required": [ "tool", "results" ]

**Comment by michaelcfanning** [comment]
btw, absolutely agree that an explicit empty results array is required to indicate 'i did my job, and there were no results'. we probably need a specific section/statement around 'how to determine that an analysis run succeeded'. consider the case where some results were produced, but we have a tool notification error. that run is not good. is it ok for tool or configuration warnings to appear, however? generally, we need an explicit indicator somewhere 'we encountered a condition that may have resulted in dropped analysis'.

**Comment by lgolding** [comment]
Yes, that is a bug in the v2 schema.

I think it happened when I was advocating for a separate schema for "results-bearing" SARIF files and "resource-only" SARIF files; then `results` would have been required for results-bearing files (except for the point you raised in this issue). Since we never adoption that separation of schemas, it was a mistake for me to put `results` in the `required` array -- even without knowing about the current issue.

**Comment by michaelcfanning** [comment]
Ok, no worries. So this will be ironed out once we converge on an updated schema in the MS SARIF SDK repo (which I will subsequently submit as a PR here)
